### PR TITLE
Fix leaking reconciler storage

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref https://github.com/kubernetes/kubernetes/issues/108483

```release-note
NONE
```

/kind cleanup
/priority important-longterm
/sig api-machinery

### CODE REVIEW

File: pkg/controlplane/reconcilers/lease.go [hunk 1]
```diff
@@ -117,13 +123,23 @@ func (s *storageLeases) RemoveLease(ip string) error {
...
+}
+
 // NewLeases creates a new etcd-based Leases implementation.
-func NewLeases(storage storage.Interface, baseKey string, leaseTime time.Duration) Leases {
+func NewLeases(config *storagebackend.ConfigForResource, baseKey string, leaseTime time.Duration) (Leases, error) {
+	leaseStorage, destroyFn, err := storagefactory.Create(*config, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error creating storage factory: %v", err)
+	}
+	var once sync.Once
```
Comment (aojea) [inline pkg/controlplane/reconcilers/lease.go:136]: why not making it part of the object?

File: pkg/controlplane/reconcilers/lease.go [hunk 2]
```diff
@@ -307,4 +323,5 @@ func (r *leaseEndpointReconciler) StopReconciling() {
 	r.reconcilingLock.Lock()
 	defer r.reconcilingLock.Unlock()
 	r.stopReconcilingCalled = true
+	r.masterLeases.Destroy()
```
Comment (aojea) [inline pkg/controlplane/reconcilers/lease.go:326]: some of the tests are dumping a lot of stack traces about grpc, can be related to this?
can we really destroy the storage?

File: pkg/controlplane/reconcilers/lease.go (hunk 1 above)
Comment (aojea) [reply pkg/controlplane/reconcilers/lease.go:136]: it is a naive question, what will be the difference, runtime allocation?

**Comment by aojea** [comment]
https://github.com/kubernetes/kubernetes/pull/109760#discussion_r863538773

same tests failed (2/2)

> k8s.io/kubernetes/test/integration/scheduler: TestPodAffinityScoring expand_more	0s
> k8s.io/kubernetes/test/integration/scheduler: TestPodAffinityScoring/pod_affinity expand_more

File: pkg/controlplane/reconcilers/lease.go (hunk 1 above)
Comment (wojtek-t) [reply pkg/controlplane/reconcilers/lease.go:136]: I don't have any strong preference - either works for me. And I don't think there is huge difference between those two.

File: pkg/controlplane/reconcilers/lease.go (hunk 2 above)
Comment (wojtek-t) [reply pkg/controlplane/reconcilers/lease.go:326]: The storage is owned by Leases (and nothing else should be accessing it).

The `stopReconcilingCalled` when set to true, nothing more should try to access the underlying masterLeases, so conceptually it is safe to destroy it.

The test that are failing are failing with timeouts:
```
panic: test timed out after 10m0s
```

I have a feeling that it's something else, but let's hold the PR until we debug it.

**Review by wojtek-t — COMMENTED** [review]
/hold

**Comment by wojtek-t** [comment]
> https://github.com/kubernetes/kubernetes/pull/109760#discussion_r863538773
>
> same tests failed (2/2)
>
>> k8s.io/kubernetes/test/integration/scheduler: TestPodAffinityScoring expand_more 0s
>> k8s.io/kubernetes/test/integration/scheduler: TestPodAffinityScoring/pod_affinity expand_more


Looking into those failures, they don't look the same.

The older one failed on shutdown - the test effectively passed (from logs)
```
    priorities_test.go:331: Pod pod1 got successfully scheduled on node in-topology-3.
```
https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler/priorities_test.go#L331

and it was already shutting down (and timed out 2s after that).

There is a number of logs like this:
```
{"level":"warn","ts":"2022-05-02T20:42:56.322Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc085a2afc0/127.0.0.1:2379","attempt":64,"error":"rpc error: code = Canceled desc = grpc: the client connection is closing"}
```
And one log like that:
```
E0502 20:42:56.336143  119235 event_broadcaster.go:262] Unable to write event: 'Patch "http://127.0.0.1:44309/apis/events.k8s.io/v1/namespaces/permit-plugins5764f134-f5cc-4a96-bbc9-5439d688c1d0/events/test-pod.16eb642340c4df37": dial tcp 127.0.0.1:44309: connect: connection refused' (may retry after sleeping)
```
But that one seems unrelated to this PR (it's for events and events are not touched by this PR).


The newer one didn't even start the test effectively - the apiserver was still initializing:
```
goroutine 209606 [select]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.WaitForWithContext({0x55f94c8, 0xc000138000}, 0xc00b0c8888, 0xbe896a?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:658 +0xe5
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.poll({0x55f94c8, 0xc000138000}, 0x60?, 0xbe7385?, 0x18?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:594 +0x9a
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.PollImmediateWithContext({0x55f94c8, 0xc000138000}, 0x11e6c85?, 0xc00ceeb5b0?, 0x40f8e7?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:526 +0x4a
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.PollImmediate(0xc00ceeb838?, 0xc018e15260?, 0x5609a30?)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:512 +0x50
k8s.io/kubernetes/test/integration/framework.startAPIServerOrDie(0xc01806f880, 0xc0105ff500, {0x55c93e0, 0xc00cbe6710})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/framework/controlplane_utils.go:279 +0x11cd
k8s.io/kubernetes/test/integration/framework.RunAnAPIServerUsingServer(...)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/framework/controlplane_utils.go:403
k8s.io/kubernetes/test/integration/util.InitTestAPIServer(0x4d40763?, {0xc0164636f0, 0x10}, {0x0, 0x0})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/util/util.go:351 +0x1fb
k8s.io/kubernetes/test/integration/scheduler.initTestSchedulerForPriorityTest(0xc0149631e0?, {0x4d40763, 0x10})
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/scheduler/priorities_test.go:68 +0x29c
k8s.io/kubernetes/test/integration/scheduler.TestPodAffinityScoring.func1(0xc0149631e0)
	/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/scheduler/priorities_test.go:298 +0xb0
testing.tRunner(0xc0149631e0, 0xc010b98fa0)
	/usr/local/go/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1486 +0x35f
```

[This initialization has 30s timeout and the test timed out much sooner than that:
https://github.com/kubernetes/kubernetes/blob/master/test/integration/framework/controlplane_utils.go#L279-L291
]

So in fact - both failures seem to me like real timeout.

My suspicion is that with scheduler tests we're actually very close to timeout as there are a lot of them...

@aojea - thoughts?

**Comment by aojea** [comment]
/test pull-kubernetes-integration 

:eyes:

**Comment by aojea** [comment]
well, I don't know but this PR seems to consistently hit errors on the integration job :smile: , 


/test pull-kubernetes-integration

File: pkg/controlplane/reconcilers/lease.go (hunk 1 above)
Comment (aojea) [reply pkg/controlplane/reconcilers/lease.go:136]: it was just curiosity

**Comment by wojtek-t** [comment]
OK - I know what is happening here - is seems that it is basically adding 2-3 seconds to the execution of each test. Will send out a fix soon.

**Comment by wojtek-t** [comment]
Unrelated flake: #107414

/test pull-kubernetes-unit

/hold cancel

**Pull request merged**